tracer/src/Datadog.Trace/Sampling/SpanSamplingRule.cs

```diff
@@ -0,0 +1,159 @@
+        /// <param name="maxPerSecond">The maximum number of spans allowed to be kept per second - <see langword="null"/> indicates that there is no limit</param>
+        public SpanSamplingRule(string serviceNameGlob, string operationNameGlob, float samplingRate = 1.0f, float? maxPerSecond = null)
+        {
+            if (string.IsNullOrWhiteSpace(serviceNameGlob))
+            {
+                throw new ArgumentException("Value cannot be null or whitespace.", nameof(serviceNameGlob));
+            }
+
+            if (string.IsNullOrWhiteSpace(operationNameGlob))
+            {
+                throw new ArgumentException("Value cannot be null or whitespace.", nameof(operationNameGlob));
+            }
+
+            SamplingRate = samplingRate;
+            MaxPerSecond = maxPerSecond;
+
+            _serviceNameRegex = ConvertGlobToRegex(serviceNameGlob);
+            _operationNameRegex = ConvertGlobToRegex(operationNameGlob);
+
+            // null/absent for MaxPerSecond indicates unlimited, which is a negative value in the limiter
+            _limiter = MaxPerSecond is null ? new SpanRateLimiter(-1) : new SpanRateLimiter((int?)MaxPerSecond);
+        }
+
+        /// <inheritdoc/>
+        public float SamplingRate { get; } = 1.0f;
+
+        /// <inheritdoc/>
+        public float? MaxPerSecond { get; } = null;
+
+        /// <summary>
+        ///     Creates <see cref="SpanSamplingRule"/>s from the supplied JSON <paramref name="configuration"/>.
+        /// </summary>
+        /// <param name="configuration">The JSON-serialized configuration.</param>
+        /// <returns><see cref="IEnumerable{T}"/> of <see cref="SpanSamplingRule"/>.</returns>
+        public static IEnumerable<SpanSamplingRule> BuildFromConfigurationString(string configuration)
+        {
+            if (string.IsNullOrWhiteSpace(configuration))
+            {
+                return Enumerable.Empty<SpanSamplingRule>();
+            }
+
+            try
+            {
+                var rules = JsonConvert.DeserializeObject<List<SpanSamplingRuleConfig>>(configuration);
+                return rules?.Select(
+                                  rule => new SpanSamplingRule(
+                                      rule.ServiceNameGlob,
+                                      rule.OperationNameGlob,
+                                      rule.SampleRate,
+                                      rule.MaxPerSecond))
+                             .ToList() ?? Enumerable.Empty<SpanSamplingRule>();
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, "Unable to parse the span sampling rule.");
+                return Enumerable.Empty<SpanSamplingRule>();
+            }
+        }
+
+        /// <inheritdoc/>
+        public bool ShouldKeep(Span span)
+        {
+            // TODO should we log here?
+            if (!IsMatch(span))
+            {
+                return false;
```

Comment:
I'm not quite sure how `ShouldKeep` will be used, but I'm a little concerned that the decision for "didn't match" and "chose not to sample" both result in a `false` return value. What do you think about making the return type `bool?`? Or perhaps we require the caller to call `IsMatch` first so we always know that the rule applies?

Reply:
I like the idea of breaking it up to try and call the `IsMatch` first and then go and making the sampling decision with some separate function

I'll try that out and see how that works with the `SpanSampler` implementation

Reply:
Split these up in 300ce7882978e93f4ec824254d64177e97799db7
To be a `IsMatch` that returns a `bool` for whether the glob patterns match and `ShouldSample` that returns `bool` for whether the sampling rate/limit are good